src/utils/queue.js

```diff
@@ -0,0 +1,42 @@
+import EventEmitter from 'eventemitter3';
+
+export default class Queue {
+    constructor(handlerTask, maxParallelTasks) {
+        this.emitter = new EventEmitter();
+        this.handlerTask = handlerTask;
+        this.maxParallelTasks = maxParallelTasks;
+        this.queue = [];
+        this.tasks = [];
+    }
+
+    pushToQueue(task) {
+        this.queue.push(task);
+        this.handlerQueue();
+    }
+
+    removeTask(filePath) {
+        this.tasks.splice(this.tasks.findIndex(taskFilePath => taskFilePath === filePath), 1);
+        this.queue.splice(this.queue.findIndex(taskFilePath => taskFilePath === filePath), 1);
+        this.handlerQueue();
+    }
+
+    onQueueEmpty(handler) {
+        this.emitter.on('queueEmpty', handler);
+    }
+
+    handlerQueue() {
+        let currentFilePath = this.queue[this.tasks.length];
+
+        if (!currentFilePath) {
+            this.emitter.emit('queueEmpty');
+            return;
+        }
+
+        if (this.tasks.length >= this.maxParallelTasks) {
+            return;
+        }
+
+        this.tasks.push(currentFilePath);
+        this.handlerTask(this, currentFilePath);
```

Comment:
It's not for sync task only. `maxParallelTasks` must work fine in other cases when we will reuse class Queue. `handlerTask` must call method `removeTask()` and it will be a signal to start handle next task in queue.

Reply:
It's a bad idea to pass responsibilities to a handler. It's not obvious at all.
If you want to handle async tasks — handler should return promise. If it doesn't then task is sync. And you can remove it from tasks list right after execution.

Reply:
removed at all